src/theory/arith/nl/cad/cdcac.cpp

```diff
@@ -54,36 +91,300 @@ const std::vector<poly::Variable>& CDCAC::getVariableOrdering() const
 std::vector<CACInterval> CDCAC::getUnsatIntervals(
     std::size_t cur_variable) const
 {
-  return {};
+  std::vector<CACInterval> res;
+  for (const auto& c : d_constraints.getConstraints())
+  {
+    const poly::Polynomial& p = std::get<0>(c);
+    poly::SignCondition sc = std::get<1>(c);
+    const Node& n = std::get<2>(c);
+
+    if (main_variable(p) != d_variableOrdering[cur_variable])
+    {
+      // Constraint is in another variable, ignore it.
+      continue;
+    }
+
+    Trace("cdcac") << "Infeasible intervals for " << p << " " << sc
+                   << " 0 over " << d_assignment << std::endl;
+    auto intervals = infeasible_regions(p, d_assignment, sc);
+    for (const auto& i : intervals)
+    {
+      Trace("cdcac") << "-> " << i << std::endl;
+      std::vector<poly::Polynomial> l, u, m, d;
+      if (!is_minus_infinity(get_lower(i))) l.emplace_back(p);
+      if (!is_plus_infinity(get_upper(i))) u.emplace_back(p);
+      m.emplace_back(p);
+      res.emplace_back(CACInterval{i, l, u, m, d, {n}});
+    }
+  }
+  cleanIntervals(res);
+  return res;
 }
 
 std::vector<poly::Polynomial> CDCAC::requiredCoefficients(
     const poly::Polynomial& p) const
 {
-  return {};
+  std::vector<poly::Polynomial> res;
+  for (long deg = degree(p); deg >= 0; --deg)
+  {
+    auto coeff = coefficient(p, deg);
+    if (lp_polynomial_is_constant(coeff.get_internal())) break;
+    res.emplace_back(coeff);
+    if (evaluate_constraint(coeff, d_assignment, poly::SignCondition::NE))
+    {
+      break;
+    }
+  }
+  return res;
 }
 
 std::vector<poly::Polynomial> CDCAC::constructCharacterization(
     std::vector<CACInterval>& intervals)
 {
-  return {};
+  Assert(!intervals.empty()) << "A covering can not be empty";
+  Trace("cdcac") << "Constructing characterization now" << std::endl;
+  std::vector<poly::Polynomial> res;
+
+  for (const auto& i : intervals)
+  {
+    Trace("cdcac") << "Considering " << i.d_interval << std::endl;
+    Trace("cdcac") << "-> " << i.d_lowerPolys << " / " << i.d_upperPolys
+                   << " and " << i.d_mainPolys << " / " << i.d_downPolys
+                   << std::endl;
+    Trace("cdcac") << "-> " << i.d_origins << std::endl;
+    for (const auto& p : i.d_downPolys)
+    {
+      // Add all polynomial from lower levels.
+      addPolynomial(res, p);
+    }
+    for (const auto& p : i.d_mainPolys)
+    {
+      Trace("cdcac") << "Discriminant of " << p << " -> " << discriminant(p)
+                     << std::endl;
+      // Add all discriminants
+      addPolynomial(res, discriminant(p));
+
+      for (const auto& q : requiredCoefficients(p))
+      {
+        // Add all required coefficients
+        Trace("cdcac") << "Coeff of " << p << " -> " << q << std::endl;
+        addPolynomial(res, q);
+      }
+      // TODO(#4814): Only add if p(s \times a) = 0 for some a <= l
```

Comment:
Can you update this reference to the project issue # and write `cvc4-projects #N`?